File: QID-617-SFEclaytonMC/SFEclaytonMC.R

# clear variables and close windows
rm(list = ls(all = TRUE))
graphics.off()

# install and load packages
libraries = c("copula")
lapply(libraries, function(x) if (!(x %in% installed.packages())) {
install.packages(x)
})
lapply(libraries, library, quietly = TRUE, character.only = TRUE)

theta = 0.79  # Set theta [0,1]

# Variables uniformly distributed:
uniclayMVD = mvdc(claytonCopula(theta), margins = c("unif", "unif"), paramMargins = list(list(min = 0, 
    max = 1), list(min = 0, max = 1)))
uniclay = rMvdc(uniclayMVD, n = 10000)

# Variables standard normally distributed:
normclayMVD = mvdc(claytonCopula(theta), margins = c("norm", "norm"), paramMargins = list(list(mean = 0, 
    sd = 1), list(mean = 0, sd = 1)))
normclay = rMvdc(normclayMVD, n = 10000)

# Set up plotting grid
plot(uniclay, main="Uniform Marginal Distribution", xlab = "", ylab = "", cex = 0.7, pch = 19, cex.axis = 1.8)
dev.new()
plot(normclay, main="Normal Marginal Distribution", xlab = "", ylab = "", cex = 0.7, pch = 19, cex.axis = 1.8)

File: QID-617-SFEclaytonMC/SFEclaytonMC.m
clear all
close all
clc

theta  = 0.79;     %dependence parameter of clayton copula
sample = 10000;    %sample size

% main computation
r   = gamrnd(theta^(-1), 1, sample, 1);
x1  = unifrnd(0, 1, sample, 1);
x2  = unifrnd(0, 1, sample, 1);
u1  = log(x1)./r;
u2  = log(x2)./r;
u   = [u1, u2];
g   = (1 - u).^(-(theta^(-1)));
rr1 = norminv(g(:, 1), 0, 1);
rr2 = norminv(g(:, 2), 0, 1);
rr  = [rr1, rr2];

% output
figure;
scatter(rr(:, 1), rr(:, 2), 'k', '.'); 
xlabel('X');
ylabel('Y');
title('Normal Marginal Distribution')
figure
scatter(g(:, 1), g(:, 2), 'k', '.');
xlabel('X');
ylabel('Y');
title('Uniform Marginal Distribution')